### DISCUSSION TITLE
Endpoints respond relatively slow

### DISCUSSION THREAD

**Discussion by Haydnspass** [fastapi-users/fastapi-users#686] [Q&A] [answered]
Hey,
Running the bare Full Example using SQL Alchemy and uvicorn (on my 2016 MacBook Pro) results in relatively slow response times.
For example, _register_ and token on login take around 300-500ms per request which I think is quite a bit, no? I know it's only a MBP but still?

Is there something that I am overlooking? 
I don't think it is the database, using bare sqlalchemy creating 1000 rows are much faster.

Thanks in advance :)

**Comment by frankie567** [comment, marked as answer]
Hi!

Actually, this is completely normal and expected. Indeed, `/register` and `/login` routes have to perform a quite expensive operation: password hashing. The point of this is to transform a plain password into a hashed string representing this password that can't be reverted to the original password. It's a crucial security mechanism to avoid passwords to be stolen by an attacker in the case they gain access to your database. More details about this here: https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html

Those hashing algorithms need to make several "pass" on the password to be robust and secure (see https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#work-factors), that's why the response time is magnitude slower than other endpoints. This is completely standard in the industry and happens in every frameworks implementing password-based authentication.

You can see the corresponding code [here](https://github.com/frankie567/fastapi-users/blob/6374cf4b7a9d3d4a6e7b2d87c2c9657687701bda/fastapi_users/user.py#L69) and [here](https://github.com/frankie567/fastapi-users/blob/6374cf4b7a9d3d4a6e7b2d87c2c9657687701bda/fastapi_users/db/base.py#L56-L64).
